pkg/kubectl/cmd/util/factory.go

```diff
@@ -173,10 +173,23 @@ func DefaultGenerators(cmdName string) map[string]kubectl.Generator {
 	return generators[cmdName]
 }
 
+// ConfigOption allows plugins to inject behavior that modifies the default client configuration
+type ConfigOption interface {
+	Flags(*pflag.FlagSet, clientcmd.ConfigOverrideFlags, *clientcmd.ClientConfigLoadingRules, *clientcmd.ConfigOverrides)
+	ClientConfig(clientcmd.ClientConfig) clientcmd.ClientConfig
+}
+
+// EmptyConfigOption provides a noop implementation of ConfigOption that plugins can embed
+type EmptyConfigOption int
+
+func (o *EmptyConfigOption) ClientConfig(c clientcmd.ClientConfig) clientcmd.ClientConfig { return c }
+func (o *EmptyConfigOption) Flags(*pflag.FlagSet, clientcmd.ConfigOverrideFlags, *clientcmd.ClientConfigLoadingRules, *clientcmd.ConfigOverrides) {
+}
+
 // NewFactory creates a factory with the default Kubernetes resources defined
 // if optionalClientConfig is nil, then flags will be bound to a new clientcmd.ClientConfig.
 // if optionalClientConfig is not nil, then this factory will make use of it.
-func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
+func NewFactory(optionalClientConfig clientcmd.ClientConfig, configOptions ...ConfigOption) *Factory {
```

Comment:
I don't think this is how we should be doing this.  If this does become a recommended way to change the kubectl factory, I think the code is more coherent if a certain kind of plugin can provide a different factory implementation.  This is actually why the factory exists: it is *the* way to specify behavior that is otherwise global to kubectl commands.  

Trying to handle this as an argument that "flavors" the return value for "the one area I want to touch" effectively means passing this list to every bit down the chain of factory creation.  This makes it easy for the one you're adding, but structurally harder for everyone following you.